GG/src/RichText/RichText.cpp

```diff
@@ -0,0 +1,301 @@
+                m_blockFactoryMap = blockFactoryMap;
+            }
+
+        private:
+            RichText* const m_owner; //!< The public control.
+            boost::shared_ptr<Font> m_font; //!< The font to use for text.
+            Clr m_color; //! < The color to use for text.
+            Flags<TextFormat> m_format; //!< Text format.
+            boost::shared_ptr< RichText::BLOCK_FACTORY_MAP > m_blockFactoryMap; //!< A map that tells us how to generate block controls from tags.
+            std::vector<BlockControl*> m_blocks; //!< The blocks generated from our content.
+            int m_padding;
+
+            // Easier access to m_blockFactoryMap
+            RichText::BLOCK_FACTORY_MAP& FactoryMap() {
+                return *m_blockFactoryMap;
+            }
+
+            // Create blocks from tags.
+            void PopulateBlocks (const std::vector< RichTextTag >& tags) {
+                // Create blocks using factories.
+                BOOST_FOREACH (const RichTextTag & tag, tags) {
+                    RichText::TAG_PARAMS params;
+                    // Extract the parameters from params_string to the tag_params map.
+                    ExtractParameters (tag.tag_params, params);
+
+                    BlockControl* block = FactoryMap() [tag.tag]->CreateFromTag (tag.tag, params, tag.content, m_font, m_color, m_format);
+                    m_owner->AttachChild (block);
+                    m_blocks.push_back (block);
+                }
+
+                DoLayout();
+            }
+
+            // Update the sizes of all blocks.
+            void DoLayout() {
+                X width = m_owner->ClientWidth() - X (m_padding);
+                Pt pos = Pt (X (m_padding), Y (m_padding));
+
+                // The contract between RichText and block controls is this:
+                // RichText tells them their width, and they determine their height.
+                BOOST_FOREACH (BlockControl * block, m_blocks) {
+                    Pt size = block->SetMaxWidth (width);
+                    block->MoveTo (pos);
+                    pos.y += size.y;
+                }
+
+                Pt size (m_owner->Width(), pos.y + Y (m_padding));
+                m_owner->Resize (size);
+            }
+    };
+
+/////////////////////////////////
+/// RichText public interface //
+///////////////////////////////
+
+    RichText::RichText (X x, Y y, X w, Y h, const std::string& str, const boost::shared_ptr<Font>& font,
+                        Clr color, Flags<TextFormat> format,
+                        Flags<WndFlag> flags) :
+        Control (x, y, w, h, flags),
+        m_self (new RichTextPrivate (this, str, font, color, format)) {
+        // Set text requires our members to be set.
+        // Therefore we must call it here, not in constructor of Private.
+        m_self->SetText (str);
+    }
+
+    RichText::~RichText() {
```

Comment:
in GG code, put the { on a new line

the one-line function can be on the same line as { and } or between their lines.